libdnf/utils/utils_internal.hpp

```diff
@@ -69,6 +70,16 @@ inline Id id_to_lowercase_id(Pool * pool, Id id_input, int create) {
     return id_to_lowercase_id(pool, name_cstring, create);
 }
 
+template <class TContainer, class T>
+inline T join(const TContainer & elements, const T & joiner) {
```

Comment:
Commit name is "Add function to accumulate strings".

1. Maybe restricting the template to always return `std::string` (and joinable to `std::string`) will be safer.
2. The template does not work items of an ordinary C array. C array does not have `begin()`, `end()`, `empty()` methods. Can be replaced with `std::begin()`, `std::end()`, and test `begin == end`.

I gave a similar function as an example (traditional approach without using `std::accumulate`) here https://github.com/rpm-software-management/libdnf/pull/1154#discussion_r590505077.

Reply:
Thanks for suggestion. I used your code with minor changes.